### PR TITLE
Update CUDA Array Interface & Enforce Numba compliance

### LINKED ISSUE #4140: Numba CUDA kernels fail when passing 0-size cupy arrays
## Reporting a bug

- [x] I am using the latest released version of Numba
- [x] I have included below a minimal working reproducer

When passing a `cupy` array of size 0 to a Numba-CUDA kernel, the kernel launch fails with
```
 Call to cuPointerGetAttribute results in CUDA_ERROR_INVALID_VALUE
```

Here is a minimal example:
```python
import cupy
from numba import cuda

@cuda.jit
def fill_with_zeros(x, N):
    i = cuda.grid(1)
    if i < N:
        x[i] = 0

def get_array_filled_with_zeros(N):
    x = cupy.ones(N)
    assert N<128  # kernel is launched with 128 threads
    fill_with_zeros[1,128](x, N)
    return x

if __name__ == '__main__':
    get_array_filled_with_zeros(10) # Passes
    get_array_filled_with_zeros(1)  # Passes
    get_array_filled_with_zeros(0)  # Fails
```

### LINKED ISSUE #4175: __cuda_array_interface__ is not compliant with the standard
**UPDATE: reporting two issues on `__cuda_array_interface__`**

First, currently it's impossible for `'strides'` to be `None`, and it is always there instead of being optional:

https://github.com/numba/numba/blob/a97dc18f3e5d69429949e557bc98ec513f81b04c/numba/cuda/cudadrv/devicearray.py#L120

If the array is C contiguous, as in most cases, this is not complying with the standard:
> The following are optional entries:
> - strides: None or (integer, ...)
>  A tuple of int (or long) representing the number of bytes to skip to access the next element at each dimension. If it is None, the array is assumed to be in C-contiguous layout.

Second, if the input is a 0-size array, the device pointer given by Numba's `__cuda_array_interface__['data'][0]` is `None`. But this leads to an error when accessing the pointer on the C/Cython side. For example, `PyLong_AsVoidPtr(None)` fails, and so it requires extra logic to handle this special case. In fact, this again breaks the spec, which requires an `int` or `long`:
> - data: (integer, boolean)
> The data is a 2-tuple. The first element data pointer as a Python int (or long). The data must be device-accessible. The second element is the read-only flag as a Python bool.

I suggest to revise the standard, as currently this edge case is not clearly defined. Relevant to #4140. By the way, CuPy uses `0` for this case (cupy/cupy#2104), which makes perfect sense at the C level.

### PR DESCRIPTION
**Pull request by leofang**
Closes #4175. Fixes #4140.

Changes (**UPDATED**): 
- `__cuda_array_interface__` is now always a Python dict (see below)
- clarified that when `strides` is missing or when `strides=None` it's meant to be used as a flag to indicate C contiguity (in line with Numpy's `__array_interface__`, see discussions in #4175 and https://github.com/numba/numba/pull/4609#discussion_r328986919)
- added that the data pointer is `0` for zero-sized arrays
- added two tests for compliance
- bumped `__cuda_array_interface__` version to 2
- updated docs

Note that I slightly modified the language in the doc to enforce `__cuda_array_interface__` to be a Python dict. This allows cleaner treatment on the C/Cython side. For example, instead of
```python
cdef object cu_arr_dict = arr.__cuda_array_interface__
```
we can do 
```python
cdef dict cu_arr_dict = arr.__cuda_array_interface__
```
potentially allowing further optimization by the compiler. While this change again aligns with `__array_interface__` in NumPy and doesn't break backward compatibility with downstream libraries AFAIK, it hasn't been discussed before so I raise it here.

Once this PR is merged, downstream consumers & library providers of `__cuda_array_interface__` should be notified.

cc: @pentschev @stuartarchibald @kkraus14 @dalcinl

### CODE REVIEW

File: numba/cuda/cudadrv/devicearray.py [hunk 1]
```diff
@@ -116,10 +116,20 @@ def __init__(self, shape, strides, dtype, stream=0, writeback=None,
 
     @property
     def __cuda_array_interface__(self):
+        if self.device_ctypes_pointer.value is not None:
+            ptr = self.device_ctypes_pointer.value
+        else:
+            ptr = 0
```
Comment (pentschev) [inline numba/cuda/cudadrv/devicearray.py]: I think we should document that `ptr` can be `0`, and what that means.

Comment (pentschev) [reply numba/cuda/cudadrv/devicearray.py]: Or rather document that the `data` may include a value `0` as first element.

**Comment by pentschev** [comment]
LGTM, I just added a comment on documenting the meaning of a `0`-valued pointer. Also, the failing CI seems unrelated, one of the tests was cancelled for no clear reason (probably timeout).

**Comment by pentschev** [comment]
Thanks for working on this @leofang!

File: numba/cuda/cudadrv/devicearray.py (hunk 1 above)
Comment (leofang) [reply numba/cuda/cudadrv/devicearray.py]: @pentschev Thanks for quick reply! How about this:
```diff
diff --git a/docs/source/cuda/cuda_array_interface.rst b/docs/source/cuda/cuda_array_interface.rst
index fa14d41..1aa3e95 100644
--- a/docs/source/cuda/cuda_array_interface.rst
+++ b/docs/source/cuda/cuda_array_interface.rst
@@ -34,8 +34,9 @@ contain the following entries:
 
 - **data**: `(integer, boolean)`
 
-    The **data** is a 2-tuple.  The first element data pointer
+    The **data** is a 2-tuple.  The first element is the data pointer
     as a Python `int` (or `long`).  The data must be device-accessible.
+    For zero-size arrays, use `0` here.
     The second element is the read-only flag as a Python `bool`.
 
     Because the user of the interface may or may not be in the same context,
```

File: numba/cuda/cudadrv/driver.py [hunk 2]
```diff
@@ -1746,8 +1746,9 @@ def get_devptr_for_active_ctx(ptr):
     pointer.
     """
     devptr = c_void_p(0)
-    attr = enums.CU_POINTER_ATTRIBUTE_DEVICE_POINTER
-    driver.cuPointerGetAttribute(byref(devptr), attr, ptr)
+    if ptr != 0:
```
Comment (sklam) [inline numba/cuda/cudadrv/driver.py]: the `ptr == 0` condition is never exercised by the test.  Can you add a test that ingest a zero-sized array from the cuda array interface?

Comment (leofang) [reply numba/cuda/cudadrv/driver.py]: OK I'm on it.

File: numba/cuda/cudadrv/devicearray.py (hunk 1 above)
Comment (leofang) [reply numba/cuda/cudadrv/devicearray.py]: @sklam comment for the diff above?

Comment (sklam) [reply numba/cuda/cudadrv/devicearray.py]: @leofang, thanks for pointing this out. I didn't pay enough attention to the discussion of making `strides` mandatory. I am a bit worried about this because numpy has this as optional with the default set to `None`. A consumer can retrieve the strides as `__cuda_array_interface__.get("strides")`, which will return `None` if the key is absent or return the actual value for it. I would prefer to leave it as optional.

The zero-sized array part looks good.

File: numba/cuda/cudadrv/driver.py (hunk 2 above)
Comment (leofang) [reply numba/cuda/cudadrv/driver.py]: @sklam Done, see 3b700f7.

Comment (leofang) [reply numba/cuda/cudadrv/driver.py]: wait, I was wrong. hold on.

Comment (leofang) [reply numba/cuda/cudadrv/driver.py]: OK @sklam I got it right in d266b6f.

File: numba/cuda/cudadrv/devicearray.py (hunk 1 above)
Comment (leofang) [reply numba/cuda/cudadrv/devicearray.py]: @sklam, the NumPy doc is actually wrong. The NumPy implementation actually always has `strides` in place, and set to `None` for C-contigus arrays:
```python
>>> import numpy as np
>>> a = np.arange(10)
>>> a.flags['C_CONTIGUOUS']
True
>>> a.__array_interface__['strides'] == None
True
>>> np.__version__
'1.17.2'
```
I can raise an issue in NumPy to ask them to fix it. But having a mandatory `strides` really helps us here.

cc: @pentschev

Comment (leofang) [reply numba/cuda/cudadrv/devicearray.py]: relevant lines in NumPy:
https://github.com/numpy/numpy/blob/7214ca4688545b432c45287195e2f46c5e418ce8/numpy/core/src/multiarray/getset.c#L239-L246

Comment (sklam) [reply numba/cuda/cudadrv/devicearray.py]: I interpret it as consumers should not rely on the "strides" to exist, but producers can choose to always fill it in.

Comment (leofang) [reply numba/cuda/cudadrv/devicearray.py]: As a consumer I'd also like to be able to retrieve it as `__cuda_array_interface__['strides']` without checking the presence of the key. We're gonna check if `strides is None` later anyway.

Comment (sklam) [reply numba/cuda/cudadrv/devicearray.py]: in python, you can do it in 1-line with `dict.get()`, i.e. `__cuda_array_interface__.get('strides')`, it doesn't raise.

Comment (leofang) [reply numba/cuda/cudadrv/devicearray.py]: > But having a mandatory `strides` really helps us here.

...Sorry for bombing you with multiple messages. The "help" is that we can ensure the array is C contiguous. @pentschev has suggested adding a `'c_contiguous'` field to the interface, but having this specified would do both jobs at once.

Like current Numba, PyTorch also always explicitly specifies `strides=(...)` even if it's C contiguous. I'd like to take an amended protocol to them and "strong-arm" them to make a fix 😄

Comment (leofang) [reply numba/cuda/cudadrv/devicearray.py]: Oh I didn't refresh and just saw your posts now.

Comment (leofang) [reply numba/cuda/cudadrv/devicearray.py]: @sklam are you suggesting a missing `strides` should be used to for representing C contiguity? I wouldn't mind doing this, but I need everyone on board. 

cc: @pentschev

Comment (sklam) [reply numba/cuda/cudadrv/devicearray.py]: Yes, the numpy doc say that the default is None. Thus implying c contiguous if strides is missing.

Comment (leofang) [reply numba/cuda/cudadrv/devicearray.py]: > Yes, the numpy doc say that the default is None. Thus implying c contiguous if strides is missing.

No @sklam, I think you misunderstood my reply. "Default is None" is different from "strides is missing". The former suggests that `'strides'` is always there (with value `None`), while the latter says literally `'strides'` is not there. 

Btw, see how easy we can parse the protocol differently? It'd be really, really helpful if we can say our intent loud and clear in the standard.

Comment (leofang) [reply numba/cuda/cudadrv/devicearray.py]: And let me refer to my earlier comment again:
https://github.com/numba/numba/pull/4609#discussion_r328772429

Comment (sklam) [reply numba/cuda/cudadrv/devicearray.py]: To me, "Default is None" means "None if not given".  I think this interpretation is common in numpy docs.

For instance, for `np.asarray`:

> order : {‘C’, ‘F’}, **optional**
> Whether to use row-major (C-style) or column-major (Fortran-style) memory representation. **Defaults to ‘C’.**

(from https://docs.scipy.org/doc/numpy/reference/generated/numpy.asarray.html?highlight=asarray#numpy.asarray :)

If `order` is not provided, default it to 'C'.

In Python doc, it also has the following for `dict.get`:

> get(key[, default])¶
> Return the value for key if key is in the dictionary, else default. **If default is not given, it defaults to None**, so that this method never raises a KeyError.

(from https://docs.python.org/3/library/stdtypes.html#dict.get)

In the array interface doc (https://docs.scipy.org/doc/numpy/reference/arrays.interface.html), it has:

> strides **(optional)**
> Either None to indicate a C-style contiguous array or a Tuple of strides which... **The default is None** which implies a C-style contiguous memory buffer...
> **Default: None (C-style contiguous)**

Comment (sklam) [reply numba/cuda/cudadrv/devicearray.py]: I see that cupy is taking the similar interpretation:

https://github.com/cupy/cupy/blob/36ecb4423762a79aea98bd1bbb40a3903d28ac18/cupy/core/core.pyx#L2748-L2755

Comment (leofang) [reply numba/cuda/cudadrv/devicearray.py]: @sklam I am not convinced. The comparison to `np.asarray` is apple to orange (it's just Python's default argument), and as I pointed out the NumPy implementation does not follow your interpretation. 

So, let me ask this instead: Do you insist on your interpretation in order to move on? I am fine either way, just wonder if you're still open to discussion or you're making a call. 

Thanks.

Comment (dalcinl) [reply numba/cuda/cudadrv/devicearray.py]: @leofang Maybe I'm missing something, but if we clarify that either missing `strides` or `strides=None` means C-contiguous, then consumer code just need to do:
```
strides = exporter.__cuda_array_interface__.get('strides')
if strides is None:
   ... # the buffer is C-contiguous
else:
    ... # handle strides if possible else raise exception
```

About NumPy, the docs are rather clear to me, `strides` may be missing. But let's look more carefully at the implementation.

The producer code always set `strides`, and use a `None` value to signal C contiguity. IOW, although `strides` is optional and may be missing, they decided to always set it.
https://github.com/numpy/numpy/blob/7214ca4688545b432c45287195e2f46c5e418ce8/numpy/core/src/multiarray/getset.c#L295-L297

The consumer code is in in `PyArray_FromInterface()`, please note the check `if (attr != NULL && attr != Py_None)`, that means that `strides` may be missing.
https://github.com/numpy/numpy/blob/7214ca4688545b432c45287195e2f46c5e418ce8/numpy/core/src/multiarray/ctors.c#L2633-L2634

IMHO, both numba and cupy should follow NumPy specs and implementation as close as possible. That means:

- In the interface spec, `strides` may be missing. If `strides` is missing or it is present and the value is `None`, that means the buffer is C-contiguous.
- The producer code always set `strides`, using `None` to flag C-contiguous arrays. We need tests to ensure that the numba producer code always sets `strides`.
- The consumer code do not assume `strides` is present, as it is optional. We need special tests to check that the consumer code can handle missing strides, maybe using a fake exporter as we did in mpi4py's testsuite.

Comment (leofang) [reply numba/cuda/cudadrv/devicearray.py]: Thanks @dalcinl for fresh thoughts! 

> clarify that either missing strides or strides=None means C-contiguous

We can do that.

> The producer code always set `strides`, ...
> The consumer code is ... `strides` may be missing.

I haven't thought about making a distinction in the protocol. I am a bit worried it'd be too lengthy, but I don't object for better clarity. 

Note that in the current PR we always set `strides` on the producer side.

Also note that it's not difficult to bring Numba and CuPy in sync. The CuPy counterpart is cupy/cupy#2491.

Comment (sklam) [reply numba/cuda/cudadrv/devicearray.py]: > clarify that either missing strides or strides=None means C-contiguous

Yes, we should do that.

> I haven't thought about making a distinction between producer and consumer in the protocol. I am a bit worried it'd be too lengthy, but I don't object for better clarity.

If you worry about it being too long, I think the documentation should focus on the consumer perspective first. 

Please also note that we are close to cutting a release. I would like to make sure this gets in by EOD. I am happy to accept this PR if we can reverse the decision on making `strides` mandatory.  I see the patch for zero-sized array critical for many downstream projects.

Comment (leofang) [reply numba/cuda/cudadrv/devicearray.py]: @sklam I'll try to work with you toward EOD! I'm on Gitter if you'd like to pin me.

How about this:
```diff
diff --git a/docs/source/cuda/cuda_array_interface.rst b/docs/source/cuda/cuda_array_interface.rst
index fa14d41..ba60472 100644
--- a/docs/source/cuda/cuda_array_interface.rst
+++ b/docs/source/cuda/cuda_array_interface.rst
@@ -49,14 +49,15 @@ contain the following entries:
     An integer for the version of the interface being exported.
     The current version is *2*.
 
-- **strides**: ``None`` or ``(integer, ...)``
 
-    If it is ``None``, the array is in C-contiguous layout. Otherwise, a tuple
-    of `int` (or `long`) is explicitly given for representing the number of
-    bytes to skip to access the next element at each dimension.
+The following are optional entries:
 
+- **strides**: ``None`` or ``(integer, ...)``
 
-The following are optional entries:
+    If **strides** is not given, or it is ``None``, the array is in
+    C-contiguous layout. Otherwise, a tuple of `int` (or `long`) is explicitly
+    given for representing the number of bytes to skip to access the next
+    element at each dimension.
 
 - **descr**
 ```
If the doc is OK, I'll proceed to modify the codes and tests. Thanks.

Comment (leofang) [reply numba/cuda/cudadrv/devicearray.py]: just updated the diff, see above

Comment (sklam) [reply numba/cuda/cudadrv/devicearray.py]: looks good

Comment (dalcinl) [reply numba/cuda/cudadrv/devicearray.py]: @leofang My English skills are quite limited, but shouldn't it read `Versions 1 and 2 of the CUDA Array Interface ...` ?

Comment (leofang) [reply numba/cuda/cudadrv/devicearray.py]: haha OK I'm on it. I followed the PR by @kkraus14

Comment (leofang) [reply numba/cuda/cudadrv/devicearray.py]: Marked as resolved via 836cff5

**Comment by sklam** [comment]
started smoke test

**Comment by leofang** [comment]
Updated the PR description to reflect the final status.

File: docs/source/cuda/cuda_array_interface.rst [hunk 3]
```diff
@@ -92,5 +94,13 @@ include:
 Differences with CUDA Array Interface (Version 0) 
 -------------------------------------------------
 
-The version 0 CUDA Array Interface did not have the optional **mask**
+The Version 0 of the CUDA Array Interface did not have the optional **mask**
```
Comment (dalcinl) [inline docs/source/cuda/cuda_array_interface.rst]: Is the initial *The* correct?

File: docs/source/cuda/cuda_array_interface.rst [hunk 4]
```diff
@@ -92,5 +94,13 @@ include:
...
 attribute to support masked arrays.
+
+
+Differences with CUDA Array Interface (Version 1)
+-------------------------------------------------
+
+The Versions 0 & 1 of the CUDA Array Interface neither clarified the
```
Comment (dalcinl) [inline docs/source/cuda/cuda_array_interface.rst]: Is the initial *The* correct?

File: docs/source/cuda/cuda_array_interface.rst (hunk 3 above)
Comment (leofang) [reply docs/source/cuda/cuda_array_interface.rst]: it's fine I think

**Comment by stuartarchibald** [comment]
#4626 passed on the farm, so by induction this passes.

**Comment by stuartarchibald** [comment]
da981a2  is a docs only fix, doesn't need to run on GPU hardware.

**Comment by pentschev** [comment]
Sorry for being unresponsive @leofang, but the proposed changes in the long discussion about strides seem sensible to me. Thanks again for working on that, and also @sklam and @dalcinl for the improvement suggestions!

**Comment by leofang** [comment]
Thank you @stuartarchibald @sklam @pentschev @dalcinl for your help in preparing this PR!

cc: @mrocklin in case you are interested.

**Comment by mrocklin** [comment]
cc'ing @kkraus14 and @madsbk.  I don't think that anything here should disrupt you, but you're more able to verify that than I am.

**Comment by mrocklin** [comment]
Thanks for the ping @leofang
